pkg/duck/utils.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package duck
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/validation"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	resolverFileName  = "/etc/resolv.conf"
+	defaultDomainName = "cluster.local"
+
+	// Number of characters to keep available just in case the prefix used in generateName
+	// exceeds the maximum allowed for k8s names.
+	generateNameSafety = 10
+)
+
+var (
+	domainName string
+	once       sync.Once
+
+	// Only allow alphanumeric, '-' or '.'.
+	validChars = regexp.MustCompile(`[^-\.a-z0-9]+`)
+)
+
+// GetClusterDomainName returns cluster's domain name or an error
+// Closes issue: https://github.com/knative/eventing/issues/714
+func GetClusterDomainName() string {
+	once.Do(func() {
+		f, err := os.Open(resolverFileName)
+		if err == nil {
+			defer f.Close()
+			domainName = getClusterDomainName(f)
+
+		} else {
+			domainName = defaultDomainName
+		}
+	})
+
+	return domainName
+}
+
+func getClusterDomainName(r io.Reader) string {
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		elements := strings.Split(scanner.Text(), " ")
+		if elements[0] != "search" {
+			continue
+		}
+		for i := 1; i < len(elements)-1; i++ {
+			if strings.HasPrefix(elements[i], "svc.") {
+				return elements[i][4:]
+			}
+		}
+	}
+	// For all abnormal cases return default domain name
+	return defaultDomainName
+}
+
+func ObjectRef(obj metav1.Object, gvk schema.GroupVersionKind) corev1.ObjectReference {
```

Comment:
I think you can use `kmeta.Accessor` here.

Reply:
https://github.com/GoogleCloudPlatform/cloud-run-events/issues/14

Reply:
I don't use this method. GONE